### ISSUE TITLE
`reportUninitializedInstanceVariable` does not report uninitialized instance variables hinted in class body

### ISSUE THREAD

**Issue by Tomaz-Vieira** [microsoft/pyright#3415]
**Describe the bug**
`reportUninitializedInstanceVariable` does not report uninitialized instance variables if they have [PEP 526](https://peps.python.org/pep-0526/#class-and-instance-variable-annotations) style type hints in the class body.

**To Reproduce**
```python
# pyright: strict
# pyright: reportUninitializedInstanceVariable=error

class MyClass:
    a: int

    def __init__(self) -> None: # no errors reported on constructor
        super().__init__() 

x = MyClass().a + 123 # no errors reported when instantiating
```

**Expected behavior**
An error (probably on `__init__`) indicating the uninitialized `self.a` 

**VS Code extension or command-line**
extension/cli version 1.1.243

**Comment by erictraut** [comment]
This is behaving as intended. In your example, the variable `a` is declared as a class variable, not an instance variable. If you want to effectively use `reportUninitializedInstanceVariable`, you should avoid declaring a class variable in the class body.

**Comment by Tomaz-Vieira** [comment]
>`a` is declared as a class variable, not an instance variable

Even if the variable declaration doesn't have an initial value? That way it doesn't exist in runtime.

PEP 526 makes it really confusing with their wording:

> In particular, the value-less notation `a: int` allows one to annotate instance variables that should be initialized in `__init__` or `__new__`. The proposed syntax is as follows:
> 
> ```python
>  class BasicStarship:
>     captain: str = 'Picard'               # instance variable with default
>     damage: int                           # instance variable without default
>     stats: ClassVar[Dict[str, int]] = {}  # class variable
> ```

**Comment by erictraut** [comment]
Yeah, PEP 526 is rather confusing and contradictory when it comes to class variables vs. instance variables. PEP 544 is a bit clearer, but it seems to contradict PEP 526 in places.

Here's how we've chosen to implement this in pyright. 

1. A variable that is declared in the class body and has a `ClassVar` annotation is considered to be a class variable only. It is considered a violation to attempt to overwrite it with an instance variable of the same name.
2. A variable that is declared in the class body without a `ClassVar` annotation is considered to be a class variable that is allowed to be overwritten by an instance variable.
3. A variable that is declared in an `__init__` method (or any other instance method) is considered an instance variable only, and it is an error if it is accessed from the class.

For protocol classes:
4. All instance variables must be declared within the class body. If no default value is assigned within the class body, a value must be assigned within the `__init__` method.

```python
from typing import ClassVar, Protocol

class A:
    v1: ClassVar[int] = 0  # Class variable only
    v2: ClassVar[int]  # Class variable only

    v3: int = 0  # Class variable that can be overwritten by instance variable
    v4: int  # Class variable that can be overwritten by instance variable

    def __init__(self):
        self.v5: int = 0  # Instance variable only
        self.v6 = 0  # Instance variable only

A().v1 = 1  # Error
A().v2 = 1  # Error
A().v3 = 1  # Allowed
A().v4 = 1  # Allowed
A().v5 = 1  # Allowed
A().v6 = 1  # Allowed

A.v1 = 1  # Allowed
A.v2 = 1  # Allowed
A.v3 = 1  # Allowed
A.v4 = 1  # Allowed
A.v5 = 1  # Error
A.v6 = 1  # Error

class B(Protocol):
    v1: ClassVar[int] = 0  # Class variable only
    v2: ClassVar[int]  # Class variable only

    v3: int = 0  # Class variable that can be overwritten by instance variable
    v4: int  # Class variable that can be overwritten by instance variable

    def __init__(self):
        self.v5: int = 0  # Error: must be declared in class body
        self.v6 = 0  # Error: must be declared in class body

class C(B):
    v1 = 1  # Class variable only
    v2 = 1  # Class variable only
    v3 = 1  # Instance variable with class variable default
    v4 = 1  # Instance variable with class variable default

    def __init__(self):
        self.v5: int = 0
        self.v6 = 0

C().v1 = 1  # Error
C().v2 = 1  # Error
C().v3 = 1  # Allowed
C().v4 = 1  # Allowed
C().v5 = 1  # Allowed
C().v6 = 1  # Allowed

C.v1 = 1  # Allowed
C.v2 = 1  # Allowed
C.v3 = 1  # Error
C.v4 = 1  # Error
C.v5 = 1  # Error
C.v6 = 1  # Error
```

Looking at the above example, there is one inconsistency in the way pyright handles protocols and non-protocols — specifically in handling `A.v3 = 1` versus `C.v3 = 1` (where `C.v3` is defined in a protocol class and `A.v3` is defined in a non-protocol class). The protocol behavior is dictated by PEP 544, but PEP 526 is silent about this behavior for non-protocol classes. I suspect that if we tried to make this behavior consistent, pyright would generate a lot of noise for existing code bases. So I'm reluctant to try to "fix" this inconsistency.

**Comment by erictraut** [comment]
Thinking about this a bit more, since `reportUninitializedInstanceVariable` is one of the strictest checks — not even on by default in strict mode, it's OK for it to be very opinionated. I think it's reasonable to expect that it would catch the class of bugs that you highlighted in your sample at the top of this thread. I've modified it to detect and report the case in your example.

**Comment by erictraut** [comment]
This is included in pyright 1.1.244, which I just published. It will also be included in the next release of pylance.

**Comment by ryan-gunderson** [comment]
What are people's thoughts on the following situation?

```
class Foo:
    x: int
    y: int

    def __init__(self):
        self.x = 5
        self.initialize_y()

    def initialize_y(self):
        self.y = 10
```

Here `y` is initialized but in a method called from within `__init__`.

**Comment by erictraut** [comment]
@ryan-gunderson, if you want to use `reportUninitializedInstanceVariable`, you'll need to initialize it directly within `__init__`. This diagnostic check is among the strictest (it's not even on in "strict" mode) and is necessarily very opinionated. I was reluctant to add it in the first place because it is so limiting and has the potential for so many false positives.
